File: java/PathSum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

    private boolean traverseTree(TreeNode root,int targetSum)
    {

        if(root == null)
            return false;
        targetSum -= root.val;
        //Considered a leaf if no left or right nodes
        if(root.left == null && root.right == null)
            {
                if(targetSum ==0)
                    return true;
                return false;
            }

        return traverseTree(root.left,targetSum) || traverseTree(root.right,targetSum);

    }

    public boolean hasPathSum(TreeNode root, int targetSum) {
        return traverseTree(root,targetSum);
    }
}

File: c++/PathSum.cpp
class Solution {
public:
    bool solve(TreeNode* root,int a){
        if(root == NULL)
            return a == 0;
        a -= root -> val;
        if(root -> left && root -> right == NULL)
            return solve(root ->left,a);
        else if (root -> right && root -> left == NULL)
            return solve(root -> right,a);
        else
            return solve(root -> left,a) || solve(root -> right,a);

    }
    bool hasPathSum(TreeNode* root, int targetSum) {
        //Either way, we require to reach the end
        //Root is a must
        //We go down recursively each node and see if we are able to reach 
targetsum
        if(root == NULL)
            return false;
        return solve(root,targetSum);


    }
};
